core/src/main/java/com/linecorp/armeria/client/HeadersUpdatingClient.java

```diff
@@ -0,0 +1,142 @@
+
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+
+/**
+ * The {@link HeadersUpdatingClient} enables the modification of request and response headers,
+ * allowing modifications before sending requests or after receiving responses.
+ *
+ * <p>Usage:
+ * <pre>{@code
+ * HeadersUpdatingClient.builder()
+ *     .requestHeaders()
+ *     .add(key, "a")
+ *     // Modifying request headers asynchronously
+ *     .addAsync(key, key -> UnmodifiableFuture.completedFuture(key + "b"))
+ *     .add(key, key -> "c") // key eventually becomes "abc"
+ *     .and()
+ *     .responseHeaders()
+ *     .add...
+ *     .and()
+ *     .newDecorator();
+ * }</pre>
+ * The builder allows the addition or modification of headers for both request and response
+ * in both synchronous and asynchronous manners.
+ */
+public final class HeadersUpdatingClient extends SimpleDecoratingHttpClient {
+    private final Map<CharSequence, Function<String, CompletableFuture<String>>> requestHeaderFunctionMap;
+    private final Map<CharSequence, Function<String, CompletableFuture<String>>> responseHeaderFunctionMap;
+
+    HeadersUpdatingClient(
+            HttpClient delegate,
+            Map<CharSequence, Function<String, CompletableFuture<String>>> requestHeaderFunctionMap,
+            Map<CharSequence, Function<String, CompletableFuture<String>>> responseHeaderFunctionMap) {
+        super(delegate);
+        this.requestHeaderFunctionMap = requestHeaderFunctionMap;
+        this.responseHeaderFunctionMap = responseHeaderFunctionMap;
+    }
+
+    /**
+     * Returns a new {@link HeadersUpdatingClientBuilder}.
+     */
+    public static HeadersUpdatingClientBuilder builder() {
+        return new HeadersUpdatingClientBuilder();
+    }
+
+    @Override
+    public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
+        final Map<CharSequence, Function<String, CompletableFuture<String>>> mergedRequestHeaders =
+                new HashMap<>(requestHeaderFunctionMap);
+        req.headers().forEach(
+                (k, v) -> mergedRequestHeaders.put(k, header -> UnmodifiableFuture.completedFuture(v)));
+
+        final List<CompletableFuture<Entry<CharSequence, String>>> reqHeadersListFuture = mergedRequestHeaders
+                .entrySet()
+                .stream()
+                .map(e -> e.getValue()
+                           .apply(ctx.defaultRequestHeaders().getLast(e.getKey()))
```

Comment:
I see, I understood this method acts more like a `RequestHeadersBuilder#setHeader` rather than `RequestHeadersBuilder#addHeader`.

I'm unsure how scalable this would be for the `RequestHeadersBuilder#addHeader` case (which appends a header value instead of just setting the entire header value). (i.e. using `HttpHeaderGetters.getAll(CharSequence)` will always return a single header value)

What do you think of just composing decorator-like structures like done in the first commit? (sorry that I'm looping back to this, I think the original style might be better in retrospect https://github.com/line/armeria/pull/5160/commits/cc1f82b87dcabf9ba964750bd64807a3158f338d)

Reply:
Thank you for the insightful comments. While reviewing my code, I realized that my implementation applies the decorating function only to the last header value, thereby overwriting all other header values.

1. Are you concerned that modifying my code to apply the decorating function to every header value, when there are multiple values for a header, might lead to performance issues?

2. If we revert to the previous method, I am also curious about the scope of APIs that should be supported.
  a. As you mentioned in the comment at https://github.com/line/armeria/pull/5160#issuecomment-1705280539, should we **exclude** the `header -> new header` function since it requires iterating over all header values and applying the function?
  b. Should we reduce the support to only `() -> Supplier<CompletableFuture<header>>`, as initially discussed in issue #5032, and cover the entire header value with a single value, similar to `setHeader`?

Reply:
@jrhee17 Thank you for the insightful comments. 😄 While reviewing my code, I realized that my implementation applies the decorating function only to the last header value, thereby overwriting all other header values.

1. Are you concerned that modifying my code to apply the decorating function to every header value, when there are multiple values for a header, might lead to performance issues?

2. If we revert to the previous method, I am also curious about the scope of APIs that should be supported.
  a. As you mentioned in the comment at https://github.com/line/armeria/pull/5160#issuecomment-1705280539, should we **exclude** the `header -> new header` function since it requires iterating over all header values and applying the function?
  b. Should we reduce the support to only `() -> Supplier<CompletableFuture<header>>`, as initially discussed in issue #5032, and cover the entire header value with a single value, similar to `setHeader`?